### PR TITLE
Ignore empty folder placeholder files on S3

### PR DESCRIPTION
**Pull request by alexjo2144**
These files are created by the Hadoop file system to indicate empty folders.

### CODE REVIEW

**Comment by alexjo2144** [comment]
Thanks, removed the config.

**Comment by alexjo2144** [comment]
@rohangarg messaged me on slack with the suggestion we move this change to https://github.com/prestosql/presto/blob/master/presto-hive/src/main/java/io/prestosql/plugin/hive/util/HiveFileIterator.java#L86 Do you have a preference @electrum ?

File: presto-hive/src/test/java/io/prestosql/plugin/hive/s3/MockAmazonS3.java [hunk 1]
```diff
@@ -125,6 +131,14 @@ public ListObjectsV2Result listObjectsV2(ListObjectsV2Request listObjectsV2Reque
             listingV2.getObjectSummaries().add(standardOne);
             listingV2.setTruncated(true);
             listingV2.setNextContinuationToken(continuationToken);
+
+            if (hasHadoopEmptyFolderObjects) {
+                S3ObjectSummary glacier = new S3ObjectSummary();
```
Comment (aalbu) [inline presto-hive/src/test/java/io/prestosql/plugin/hive/s3/MockAmazonS3.java:136]: Probably copy/pasta, but can you use a different variable name?

Comment (alexjo2144) [reply presto-hive/src/test/java/io/prestosql/plugin/hive/s3/MockAmazonS3.java]: Thanks for catching that

**Comment by rohangarg** [comment]
>messaged me on slack with the suggestion we move this change to https://github.com/prestosql/presto/blob/master/presto-hive/src/main/java/io/prestosql/plugin/hive/util/HiveFileIterator.java#L86 Do you have a preference?

One benefit of the split generation change approach could be that it handles custom `S3FileSystemType` as well..

**Comment by pettyjamesm** [comment]
> @rohangarg messaged me on slack with the suggestion we move this change to https://github.com/prestosql/presto/blob/master/presto-hive/src/main/java/io/prestosql/plugin/hive/util/HiveFileIterator.java#L86 Do you have a preference @electrum ?

Not electrum here, but I think it would be a good idea not to `.filter` those out here, but use it to set the `isdir` property [here instead](https://github.com/prestosql/presto/pull/4552/files#diff-bbb5f38864bd0c9c40e3e6dbcaf8b8ddL580).

That way it can be up to the caller to decide whether they want to filter those out or not (maybe via Hadoop's PathFilter API), but doing this directly for now in HiveFileIterator like you mentioned makes sense.

File: presto-hive/src/main/java/io/prestosql/plugin/hive/s3/PrestoS3FileSystem.java [hunk 2]
```diff
@@ -591,6 +592,11 @@ private static boolean isGlacierObject(S3ObjectSummary object)
         return Glacier.toString().equals(object.getStorageClass());
     }
 
+    private static boolean isHadoopEmptyFolderObject(S3ObjectSummary object)
```
Comment (findepi) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/s3/PrestoS3FileSystem.java:595]:
```suggestion
    private static boolean isHadoopFolderMarker(S3ObjectSummary object)
```

File: presto-hive/src/test/java/io/prestosql/plugin/hive/s3/MockAmazonS3.java [hunk 3]
```diff
@@ -39,6 +39,7 @@
     private GetObjectMetadataRequest getObjectMetadataRequest;
     private CannedAccessControlList acl;
     private boolean hasGlacierObjects;
+    private boolean hasHadoopEmptyFolderObjects;
```
Comment (findepi) [inline presto-hive/src/test/java/io/prestosql/plugin/hive/s3/MockAmazonS3.java:42]:
```suggestion
    private boolean hasHadoopFolderMarkers;
```

File: presto-hive/src/test/java/io/prestosql/plugin/hive/s3/TestPrestoS3FileSystem.java [hunk 4]
```diff
@@ -570,6 +570,22 @@ private static void assertSkipGlacierObjects(boolean skipGlacierObjects)
         }
     }
 
+    @Test
+    public void testSkipHadoopEmptyFolderObjectsEnabled()
```
Comment (findepi) [inline presto-hive/src/test/java/io/prestosql/plugin/hive/s3/TestPrestoS3FileSystem.java:574]: rename appropriately

**Comment by findepi** [comment]
> I think it would be a good idea not to `.filter` those out here, but use it to set the `isdir` property [here instead](https://github.com/prestosql/presto/pull/4552/files#diff-bbb5f38864bd0c9c40e3e6dbcaf8b8ddL580).

@pettyjamesm the `xxx_$folder$` itself is not a directory.  It is supposed not to exist and to cause `xxx` be treated as a folder.
I'd rather ignore them for now and reiterate when we see a need for that. This would help us provide better test cases too.

**Comment by findepi** [comment]
A follow-up: https://github.com/prestosql/presto/issues/4566

**Comment by alexjo2144** [comment]
Renamed, thanks @findepi

**Comment by findepi** [comment]
A note on filtering in HiveFileIterator -- ive does not ignore those empty files when running on hdfs://
so if we do ignore them, this may lead to some problems for end users.
This is file system specific so let's treat this on file system level.
I've created a follow up for Hive w/caching -- https://github.com/prestosql/presto/issues/4566

**Comment by findepi** [comment]
Merged, thanks!

**Pull request merged**